Add helper methods to compare `WebCore::IPAddress`es

whsieh opened the pull request:
#### 26acda475c8a24f288048011d53aa52f74a84e32
<pre>
Add helper methods to compare `WebCore::IPAddress`es
<a href="https://bugs.webkit.org/show_bug.cgi?id=248726">https://bugs.webkit.org/show_bug.cgi?id=248726</a>
rdar://99360259

Reviewed by Tim Horton.

Add a helper method to compare two `WebCore::IPAddress`-es, and use it to implement some comparison
operators. Also, add an additions hook that allows us to adjust newly created `NSURLSession`s.

Test: IPAddressTests.CompareIPAddresses

* Source/WebCore/platform/network/DNS.h:
(WebCore::IPAddress::compare const):
(WebCore::IPAddress::operator&lt; const):
(WebCore::IPAddress::operator&gt; const):
(WebCore::IPAddress::operator== const):
(WebCore::IPAddress::operator!= const):

Add the `compare()` method, and use it to implement several binary comparison operators.

* Source/WebKit/NetworkProcess/cocoa/NetworkSessionCocoa.mm:
(WebKit::applyAdditionalSettings):
(WebKit::createURLSession):

Factor out logic to create a new `NSURLSession` into a separate helper method, and call the
additions hook (`applyAdditionalSettings`) from within this new helper.

(WebKit::SessionWrapper::initialize):
(WebKit::NetworkSessionCocoa::clearCredentials):
* Tools/TestWebKitAPI/Tests/WebCore/IPAddressTests.cpp:
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/257443@main">https://commits.webkit.org/257443@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6dc577b062ab79c0c52c9b7ef6676787f766a67c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/98987 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/8193 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/32144 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/108394 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/168646 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/102949 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/8744 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/85540 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/91511 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/106360 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/104704 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/6619 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/90206 "Passed tests") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/33648 "Found 1 new test failure: media/video-as-img-output-pts.html (failure)") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/88441 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/21551 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/76502 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/2095 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/23067 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/1999 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/45463 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5126 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/6962 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/42541 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/3413 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

hortont424 commented on Source/WebCore/platform/network/DNS.h:
```diff
@@ -67,6 +67,37 @@ class IPAddress {
+        };
+
+        if (isIPv4() && other.isIPv4())
+            return comparisonResult(memcmp(&ipv4Address(), &other.ipv4Address(), 4));
```
Is there any way to avoid this hardcoded size (I hope?)

whsieh replied:
Yes! I think `sizeof(struct in_addr)` and `sizeof(struct in6_addr)` would be more explicit.

hortont424 commented on Source/WebCore/platform/network/DNS.h:
```diff
@@ -67,6 +67,37 @@ class IPAddress {
+            return comparisonResult(memcmp(&ipv4Address(), &other.ipv4Address(), 4));
+
+        if (isIPv6() && other.isIPv6())
+            return comparisonResult(memcmp(&ipv6Address(), &other.ipv6Address(), 16));
```
Ditto

whsieh committed:
Add helper methods to compare `WebCore::IPAddress`es
https://bugs.webkit.org/show_bug.cgi?id=248726
rdar://99360259

Reviewed by Tim Horton.

Add a helper method to compare two `WebCore::IPAddress`-es, and use it to implement some comparison
operators. Also, add an additions hook that allows us to adjust newly created `NSURLSession`s.

Test: IPAddressTests.CompareIPAddresses

* Source/WebCore/platform/network/DNS.h:
(WebCore::IPAddress::compare const):
(WebCore::IPAddress::operator< const):
(WebCore::IPAddress::operator> const):
(WebCore::IPAddress::operator== const):
(WebCore::IPAddress::operator!= const):

Add the `compare()` method, and use it to implement several binary comparison operators.

* Source/WebKit/NetworkProcess/cocoa/NetworkSessionCocoa.mm:
(WebKit::applyAdditionalSettings):
(WebKit::createURLSession):

Factor out logic to create a new `NSURLSession` into a separate helper method, and call the
additions hook (`applyAdditionalSettings`) from within this new helper.

(WebKit::SessionWrapper::initialize):
(WebKit::NetworkSessionCocoa::clearCredentials):
* Tools/TestWebKitAPI/Tests/WebCore/IPAddressTests.cpp:
(TestWebKitAPI::TEST):

Canonical link: https://commits.webkit.org/257443@main
